src/compiler/args.rs

```diff
@@ -437,82 +430,63 @@ where
         ArgsIter {
             arguments: arguments,
             arg_info: arg_info,
+            phantom: PhantomData,
         }
     }
 }
 
-#[derive(PartialEq, Clone, Debug)]
-pub struct ArgumentItem<T> {
-    pub arg: Argument,
-    pub data: Option<T>,
-}
-
-impl<I, T, S, U> Iterator for ArgsIter<I, T, S>
+impl<I, T, S> Iterator for ArgsIter<I, T, S>
 where
     I: Iterator<Item = OsString>,
-    T: 'static + EmbedsArgInfo<ExtraData = U>,
-    S: SearchableArgInfo<Info = T>,
+    T: ArgumentValue,
+    S: SearchableArgInfo<T>,
 {
-    type Item = ArgumentItem<T::ExtraData>;
+    type Item = ArgResult<Argument<T>>;
 
     fn next(&mut self) -> Option<Self::Item> {
         if let Some(arg) = self.arguments.next() {
             let s = arg.to_string_lossy();
             let arguments = &mut self.arguments;
-            match self.arg_info.search(&s[..]) {
-                Some(ref i) => {
-                    Some(ArgumentItem {
-                        arg: i.get_arg_info().process(&s[..], || arguments.next()),
-                        data: Some(i.get_extra()),
-                    })
+            Some(match self.arg_info.search(&s[..]) {
+                Some(i) => {
+                    i.clone().process(&s[..], || arguments.next())
                 }
                 None => {
-                    Some(ArgumentItem {
-                        arg: if s.starts_with("-") {
-                            Argument::UnknownFlag(arg.clone())
-                        } else {
-                            Argument::Raw(arg.clone())
-                        },
-                        data: None,
+                    Ok(if s.starts_with("-") {
+                        Argument::UnknownFlag(arg.clone())
+                    } else {
+                        Argument::Raw(arg.clone())
                     })
                 }
-            }
+            })
         } else {
             None
         }
     }
 }
 
 /// Helper macro used to define ArgInfo::Flag's.
+/// Variant is an enum variant, e.g. enum ArgType { Variant(()) }
 ///     flag!("-foo")
```

Comment:
You removed the single argument macro rule below, so the comment here should be removed as well.